packages/Manager/src/test/kotlin/nz/govt/eop/consumers/WaterAllocationConsumerTest.kt

```diff
@@ -0,0 +1,76 @@
+package nz.govt.eop.consumers
+
+import io.kotest.matchers.shouldBe
+import java.time.Instant
+import java.time.temporal.ChronoUnit
+import nz.govt.eop.messages.WaterAllocationMessage
+import nz.govt.eop.si.jooq.tables.WaterAllocations.Companion.WATER_ALLOCATIONS
+import nz.govt.eop.si.jooq.tables.records.WaterAllocationsRecord
+import org.jooq.DSLContext
+import org.jooq.Result
+import org.junit.jupiter.api.Test
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.boot.test.context.SpringBootTest
+import org.springframework.kafka.core.KafkaTemplate
+import org.springframework.kafka.test.context.EmbeddedKafka
+import org.springframework.test.context.ActiveProfiles
+
+@ActiveProfiles("test")
+@SpringBootTest
+@EmbeddedKafka(
+    partitions = 1,
+    bootstrapServersProperty = "spring.kafka.bootstrap-servers",
+    topics = [WATER_ALLOCATION_TOPIC_NAME])
+class WaterAllocationConsumerTest(
+    @Autowired val template: KafkaTemplate<String, WaterAllocationMessage>,
+    @Autowired val context: DSLContext
+) {
+  @Test
+  fun `Should create water allocations if they don't exist`() {
+    val message = WaterAllocationMessage("area-id-create", 100, "ingest-id", Instant.now())
+    template.send(WATER_ALLOCATION_TOPIC_NAME, message.areaId, message)
+
+    val records = fetchWaterAllocations(message.areaId)
+    records.size.shouldBe(1)
+    val record = records.first()
+    record.amount.shouldBe(message.amount)
+    record.lastUpdatedIngestId.shouldBe(message.ingestId)
+  }
+
+  @Test
+  fun `Should update water allocations if they exist`() {
+    val firstMessage = WaterAllocationMessage("area-id-update", 100, "ingest-id-1", Instant.now())
+    template.send(WATER_ALLOCATION_TOPIC_NAME, firstMessage.areaId, firstMessage)
+
+    val secondMessage = WaterAllocationMessage("area-id-update", 200, "ingest-id-2", Instant.now())
+    template.send(WATER_ALLOCATION_TOPIC_NAME, secondMessage.areaId, secondMessage)
+
+    val records = fetchWaterAllocations(firstMessage.areaId)
+    records.size.shouldBe(1)
+    val record = records.first()
+    record.amount.shouldBe(secondMessage.amount)
+    record.lastUpdatedIngestId.shouldBe(secondMessage.ingestId)
```

Comment:
Testing looks good, 

small nit: (might be something for us to discuss in general)

I'd tend to write it this way ... 

    // GIVEN
    val firstMessage = WaterAllocationMessage("area-id-update", 100, "ingest-id-1", Instant.now())
    val secondMessage = WaterAllocationMessage("area-id-update", 200, "ingest-id-2", Instant.now())

    // WHEN
    template.send(WATER_ALLOCATION_TOPIC_NAME, firstMessage.areaId, firstMessage)
    template.send(WATER_ALLOCATION_TOPIC_NAME, secondMessage.areaId, secondMessage)

    // THEN
    val records = fetchWaterAllocations(firstMessage.areaId)
    records.size.shouldBe(1)
    val record = records.first()
    record.amount.shouldBe(secondMessage.amount)
    record.lastUpdatedIngestId.shouldBe(secondMessage.ingestId)

The // GIVEN / WHEN / THEN comments are unimportant, but keeping all data setup at the start, then executing the code, then verifying ... your version interleaves setup / executing. 

I know its not do-able 100% of the time

Reply:
Happy to stick with that convention, I'll update the tests.